File: lg_rooms.py
# -*- coding: utf-8 -*-
"""
Created on Sat Mar 20 20:54:26 2021

@author: dromp_50a1bpc
"""
import enum

gridSize = 2
board = [
    [2,2],
    [1,1]
]
grange = range(gridSize)

class Dir(enum.Enum):
    UP = 0
    DOWN = 2
    LEFT = 1
    RIGHT = 3
    
class Node:
    def __init__(self, row, col, solNum):
        self.neighbors = [None]*4
        self.solNum = solNum
        self.row = row
        self.col = col
    
    def is_solved(self):
        return self.calc_curr_num() == self.solNum
    
    def is_valid(self, d):
        return self.calc_curr_num() - self.get_cnt(d) >= self.solNum

    def get_cnt(self, d):
        if self.neighbors[d]:
            return self.neighbors[d].get_cnt(d)+1
        return 0
    
    def calc_curr_num(self):
        return self.get_cnt(0) + self.get_cnt(1) + self.get_cnt(2) + self.get_cnt(3)
        
    def get_node_rank(self):
        return self.solNum + sum(1 for n in self.neighbors if n)
    
def build_nodes(b):
    node_b = [[None]*gridSize for _ in grange]
    for r in grange:
        for c in grange:
            node_b[r][c] = Node(r,c,board[r][c])
            if(r > 0):
                node_b[r][c].neighbors[Dir.UP.value] = node_b[r-1][c]
                node_b[r-1][c].neighbors[Dir.DOWN.value] = node_b[r][c]
            if(c > 0):
                node_b[r][c].neighbors[Dir.LEFT.value] = node_b[r][c-1]
                node_b[r][c-1].neighbors[Dir.RIGHT.value] = node_b[r][c]
    return node_b

def get_least_rank(ng):
    min_rn = ng[0][0]
    for r in grange:
        for c in grange:
            if ng[r][c].is_solved(): continue
            if ng[r][c].get_node_rank() < min_rn.get_node_rank():
                min_rn = ng[r][c]
    return min_rn

def is_board_solved(ng):
    for r in ng:
        for node in r:
            if not node.is_solved(): return False
    return True
    
    
def solve_rec(ng):
    if is_board_solved(ng): return (True, None)
    sn = get_least_rank(ng)
    print(sn)
    any_valid_move = False
    
    for i in range(4):
        if sn.neighbors[i] and sn.is_valid(i):
            opp = i - 2
            if opp < 0:
                opp += 4
            sn.neighbors[i].neighbors[opp] = None
            sn.neighbors[opp] = None
            any_valid_move = True
            r = solve_rec(ng)
            if r[0]: return r
    return (False, any_valid_move)
    
    
ng = build_nodes(board)
sn = solve_rec(ng)
print(is_board_solved(ng))

File: lg_rooms.java
public class lg_rooms{
    int gridSize = 5;
    int[][] board = {
        {1,1,1,1,1},
        {1,1,1,1,1},
        {1,1,1,1,1},
        {1,1,1,1,1},
        {1,1,1,1,1}
    };
    Board b;

    lg_rooms(){
        b = new Board(gridSize, board);
    }

    public static void main(String[] args){
    }
}

class Node{
    Node[] neighbors = new Node[4];
    int sol;
    int row, col;
    Node(int row, int col, int sol){
        this.sol = sol;
        this.row = row;
        this.col = col;
    }

    int getCount(Dir d){
        if(neighbors[d.val] != null)return neighbors[d.val].getCount(d) + 1;
        return 0;
    }

    int getCurNum(){
        return getCount(Dir.UP) + getCount(Dir.DOWN) + getCount(Dir.LEFT) + getCount(Dir.RIGHT);
    }

    boolean isSolved(){
        if(sol == 0) return true;
        return getCurNum() == sol;
    }

    boolean isValid(Dir d){
        if(d != null) return getCurNum() - getCount(d) >= sol;
        else return getCurNum() >= sol;
    }

    int getNeighborCount(){
        int cnt = 0;
        for(Node n : neighbors) if(n != null) cnt++;
        return cnt;
    }

    int getNodeRank(){
        return sol + getNeighborCount();
    }

}

class Board{
    int gridSize;
    int[][] board;
    Node[][] ng;
    boolean solved = false;

    Board(int gridSize, int[][] board){
        this.gridSize = gridSize;
        this.board = board;
        ng = new Node[gridSize][gridSize];

        buildBoard();
    }

    void buildBoard(){
        for(int r = 0; r < gridSize; r++){
            for(int c = 0; c < gridSize; c++){
                ng[r][c] = new Node(r, c, board[r][c]);
                if(r > 0){
                    ng[r][c].neighbors[Dir.UP.val] = ng[r - 1][c];
                    ng[r - 1][c].neighbors[Dir.DOWN.val] = ng[r][c];
                }
                if(c > 0){
                    ng[r][c].neighbors[Dir.LEFT.val] = ng[r][c - 1];
                    ng[r][c - 1].neighbors[Dir.RIGHT.val] = ng[r][c];
                }
            }
        }
    }

    Node getLeastRank(){
        Node min = ng[0][0];
        for(int r = 0; r < gridSize; r++){
            for(int c = 0; c < gridSize; c++){
                if(ng[r][c].isSolved()) continue;
                if(ng[r][c].getNodeRank() < min.getNodeRank()) min = ng[r][c];
            }
        }
        return min;
    }

    boolean isSolved(){
        for(Node[] r : ng){
            for(Node n : r) if(!n.isSolved()) return false;
        }
        return true;
    }
}

enum Dir{
    UP(0),
    DOWN(2),
    LEFT(1),
    RIGHT(3);

    int val;

    Dir(int val){
        this.val = val;
    }
}